### PR TITLE
Fix formatting in docs

### PR DESCRIPTION
**Pull request by AlexanderSher**
Unifies writing of types in docs, removes `StringBuilder` usage in the middle

### CODE REVIEW

File: src/AutoRest.CSharp/LowLevel/Generation/DpgClientWriter.cs [hunk 1]
```diff
@@ -603,19 +603,19 @@ private void WriteSampleRefsIfNecessary(MethodSignature methodSignature, bool is
                 return;
 
             var sampleProvider = _library.GetSampleForClient(_client);
-            // do not write this part when there is no sample provider
+            // do not write this part when there is no sample providerPara
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/LowLevel/Generation/DpgClientWriter.cs:606]: typo? maybe we need to revert this change

File: src/AutoRest.CSharp/Mgmt/Output/Resource.cs [hunk 2]
```diff
@@ -484,7 +484,7 @@ protected virtual Method BuildCreateResourceIdentifierMethod()
             var signature = new MethodSignature(
                 Name: "CreateResourceIdentifier",
                 null,
-                Description: $"Generate the resource identifier of a <see cref=\"{Type.ToStringForDocs()}\"/> instance.",
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Mgmt/Output/Resource.cs:487]: I assume this method `CSharpType.ToStringForDocs` is no longer needed after this change? Could we remove it?

File: src/AutoRest.CSharp/LowLevel/Generation/DpgClientWriter.cs (hunk 1 above)
Comment (AlexanderSher) [reply src/AutoRest.CSharp/LowLevel/Generation/DpgClientWriter.cs:606]: yep

File: src/AutoRest.CSharp/Mgmt/Output/Resource.cs (hunk 2 above)
Comment (AlexanderSher) [reply src/AutoRest.CSharp/Mgmt/Output/Resource.cs:487]: Done

File: src/AutoRest.CSharp/Common/Generation/Writers/DocumentationWriterExtensions.cs [hunk 3]
```diff
@@ -68,6 +70,48 @@ public static CodeWriter WriteXmlDocumentationReturns(this CodeWriter writer, Fo
...
+            {
+                AppendTypeWithShortNames(parameter.Type, sb);
+                sb.Append(",");
+            }
+
+            sb.Remove(sb.Length - 1, 1);
+            sb.Append(")");
+
+            memberId = sb.ToString();
+            return writer.LineRaw($"/// <include file=\"{filename}\" path=\"doc/members/member[@name='{memberId}']/*\" />");
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Generation/Writers/DocumentationWriterExtensions.cs:90]: does this count as "writing documentations"? should we have a check if the writer is in the mode of writing documents?

**Review by ArcturusZhang — APPROVED** [review]
LGTM

File: src/AutoRest.CSharp/Common/Generation/Writers/DocumentationWriterExtensions.cs (hunk 3 above)
Comment (AlexanderSher) [reply src/AutoRest.CSharp/Common/Generation/Writers/DocumentationWriterExtensions.cs:90]: No, we lie here to the code writer on purpose. `include` doesn't follow our type formatting rules, hence a separate method that doesn't call `AppendXmlDocumentation` and the string is fully constructed and added via `LineRaw`.

File: test/TestProjects/Models-TypeSpec/src/Generated/ModelsTypeSpecClient.cs [hunk 4]
```diff
@@ -1022,7 +1022,7 @@ public virtual Response<OutputModel> FixedFloatEnum(FixedFloatEnum input, Cancel
         /// </item>
         /// <item>
         /// <description>
-        /// Please try the simpler <see cref="FixedFloatEnumAsync(FixedFloatEnum,CancellationToken)"/> convenience overload with strongly typed models first.
+        /// Please try the simpler <see cref="FixedFloatEnumAsync(Models.FixedFloatEnum,CancellationToken)"/> convenience overload with strongly typed models first.
```
Comment (AlexanderSher) [inline test/TestProjects/Models-TypeSpec/src/Generated/ModelsTypeSpecClient.cs:1025]: `Models` namespace is needed here because there is method with the same name - `FixedFloatEnum`.

**Review by archerzz — APPROVED** [review]

**Review by m-nash — CHANGES_REQUESTED** [review]
This is a conflict with https://github.com/Azure/autorest.csharp/pull/3891.  Lets wait for this one to go in first.

**Comment by AlexanderSher** [comment]
https://github.com/Azure/azure-sdk-for-net/pull/40091

**Comment by AlexanderSher** [comment]
https://github.com/Azure/azure-sdk-for-net/pull/40126

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs [hunk 5]
```diff
@@ -102,7 +102,7 @@ public CSharpType(TypeProvider implementation, string ns, string name, bool isVa
         public bool IsLiteral { get; init; }
         public Constant? Literal { get; init; }
         public bool IsUnion { get; }
-        public CSharpType[] UnionItemTypes { get; } = Array.Empty<CSharpType>();
+        public IReadOnlyList<CSharpType> UnionItemTypes { get; } = Array.Empty<CSharpType>();
```
[the remark this answers is not in the document]
Comment (AlexanderSher) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:105]: We have mutable `CSharpType[] Arguments` as a relic of the past when arrays have been used everywhere. Now we use read-only types for public properties that shouldn't be changed. Ideally, `Arguments` should be `IReadOnlyList<CSharpType>` as well, cause changing the list of generic types would cause an error

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs [hunk 6]
```diff
@@ -152,6 +152,11 @@ private bool ArgumentsEquals(Type[] genericArguments)
 
         public override int GetHashCode() => HashCode.Combine(_implementation, _type, ((System.Collections.IStructuralEquatable)Arguments).GetHashCode(EqualityComparer<CSharpType>.Default));
 
+        public CSharpType GetGenericTypeDefinition()
```
Comment (m-nash) [inline src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:155]: Can you add a unit test which does what we discussed in chat?

CSharpType.Arguments should do what System.Type.GetGenericArguments() does
csharpType.GetGenericTypeDefinition().Arguments should do what type.GetGenericTypeDefinition().GetGenericArguments()

Then validate this with `new CSharpType(typeof(List<>), typeof(string))` and `new CSharpType(typeof<List<string>>)`

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (m-nash) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:105]: Can we add a work item to fix `Arguments` in that case we can tackle in a separate PR.

[the remark this answers is not in the document]
Comment (AlexanderSher) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:105]: https://github.com/Azure/autorest.csharp/issues/3958

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs (hunk 6 above)
Comment (AlexanderSher) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs:155]: Done

File: src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs [hunk 7]
```diff
@@ -152,6 +152,11 @@ private bool ArgumentsEquals(Type[] genericArguments)
...
+            => _type is null
+                ? new(_implementation!, Namespace, Name, IsValueType, IsEnum, IsNullable, Array.Empty<CSharpType>())
+                : new(_type, IsNullable, _type.GetGenericArguments().Select(p => new CSharpType(p)).ToArray());
```
[the remark this answers is not in the document]
Comment (AlexanderSher) [reply src/AutoRest.CSharp/Common/Generation/Types/CSharpType.cs]: Fixed per discussion offline

**Review by m-nash — APPROVED** [review]
Very nice change thank you for this clean up!
